toolchain/language_server/handle_document_symbol.cpp

```diff
@@ -0,0 +1,96 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/base/shared_value_stores.h"
+#include "toolchain/diagnostics/null_diagnostics.h"
+#include "toolchain/language_server/context.h"
+#include "toolchain/language_server/handler_registry.h"
+#include "toolchain/lex/lex.h"
+#include "toolchain/parse/node_kind.h"
+#include "toolchain/parse/parse.h"
+#include "toolchain/parse/tree_and_subtrees.h"
+#include "toolchain/source/source_buffer.h"
+
+namespace Carbon::LanguageServer {
+
+// Returns the text of first child of kind Parse::NodeKind::IdentifierName.
+static auto GetIdentifierName(const SharedValueStores& value_stores,
+                              const Lex::TokenizedBuffer& tokens,
+                              const Parse::TreeAndSubtrees& p,
+                              Parse::NodeId node)
+    -> std::optional<llvm::StringRef> {
+  for (auto ch : p.children(node)) {
+    if (p.tree().node_kind(ch) == Parse::NodeKind::IdentifierName) {
+      auto token = p.tree().node_token(ch);
+      if (tokens.GetKind(token) == Lex::TokenKind::Identifier) {
+        return value_stores.identifiers().Get(tokens.GetIdentifier(token));
+      }
+    }
+  }
+  return std::nullopt;
+}
+
+// Provides information about document symbols.
+static auto HandleDocumentSymbol(
+    Context& context, const clang::clangd::DocumentSymbolParams& params,
+    llvm::function_ref<
+        void(llvm::Expected<std::vector<clang::clangd::DocumentSymbol>>)>
+        on_done) -> void {
+  SharedValueStores value_stores;
+  llvm::vfs::InMemoryFileSystem vfs;
+  auto lookup = context.files().Lookup(params.textDocument.uri.file());
+  CARBON_CHECK(lookup);
+  vfs.addFile(lookup.key(), /*mtime=*/0,
+              llvm::MemoryBuffer::getMemBufferCopy(lookup.value()));
+
+  auto buf =
+      SourceBuffer::MakeFromFile(vfs, lookup.key(), NullDiagnosticConsumer());
+  auto lexed = Lex::Lex(value_stores, *buf, NullDiagnosticConsumer());
+  auto parsed = Parse::Parse(lexed, NullDiagnosticConsumer(), nullptr);
```

Comment:
Super minor, but I feel like naming these after what they are more consistently would be an improvement: `source`, `tokens`, and `parse_tree`? (or just `tree` for the last if you like)

Reply:
tree (given tree_and_subtrees)